### ISSUE TITLE
Add validators to Forms verifying basic data types

### ISSUE THREAD

**Issue by vsavkin** [angular/angular#2962]
Data types: email, url, number, date, time, week, month.

**Comment by naomiblack** [comment]
caitp has started on this but needs someone to pair with on design decisions vs how forms work in angular 1.

**Comment by IgorMinar** [comment]
Is there a PR? @caitp can you please provide update?

**Comment by rafacm** [comment]
Is this feature being worked on? Otherwise I would like to have a go at it. I would need some pointers in regards to whether "design decisions vs how forms work in angular 1" were made or not. Thanks!
/cc @caitp @naomiblack @IgorMinar

**Comment by mhevery** [comment]
This is obsolete, and no longer relevant / actionable. To keep our issues clean, we are aggressively closing them.

**Comment by pkozlowski-opensource** [comment]
Moved over from #11644 investigation info regarding validity of number fields (thnx @fredgate):

I analyzed the problem, also I see where it comes from.

The code in the NumberValueAccessor directive seems good :
`this.onChange = (value) => { fn(value == '' ? null : parseFloat(value)); };`
as parseFloat return NaN if the string parameter does not represent a number.

However the problem is on this line.
Remember that the NumberValueAccessor directive is associated to selectors like input[type=number]. And the browser applies some rules to this element according to the WHATWG. In the [specifications](https://html.spec.whatwg.org/multipage/forms.html#number-state-(type=number)), there is a sanitization algorithm saying that for input number field's the browser is supposed to set the value to an empty string if the input isn't a valid floating point number :

> The value sanitization algorithm is as follows: If the value of the element is not a valid floating-point number, then set it to the empty string instead.

So the value of an number input field is null when the user enters nothing, or when he enters a non numeric string. So when the NumberValueAccessor calls parseFloat it is always with a numeric string : so it never produces the NaN value.

**Comment by fredgate** [comment]
Following of #11644.
The problem with `NumberValueAccessor` is that it produces `null` value in this two cases : the user entered nothing, or the user entered a non numeric text.
So if we want to write a validator to checks that user entered a numeric value, we need to write a function that accepts an `AbstractControl` argument. But we can not determine from the control and its value if the user entered nothing (which is valid for a optional field) or entered a non numeric text (which is invalid and should produces an error we can indicate to the user).

So how a numeric validator could identify entry error in a control ? I see two options ways.

1. The validator could determine the state of the input field. For this, the validator could need to have access to the native input element, which is not the case. For the moment it has only a reference to an `AbstractControl`.
2. Change the `NumberValueAccessor` so that it returns `null` if the user enters nothing, but returns `NaN` if user enters a non numeric string. This can be easily done by adding a condition in the `registerOnChange` function. After all, why is it that the control value should be different from what it contains ? And `NumberValueAccessor` will have a consistent behavior if later it had another selector that associate it with a input of type text.

**Comment by manklu** [comment]
In my opinion it's crucial to differentiate between no input and invalid input.

**Comment by fredgate** [comment]
Where is this problem?
It is still very annoying in a final version (now 2.2), to not be able to validate the input of a number in a text field...

But actually it is not possible because a validator has only a reference to an AbstractControl. For input type=number, if the user enter a non numeric string, the value of the AbstractControl will be null. The validator can not determine if the user did not enter anything, or entered a invalid value.
So, how to solve this. I see 3 possibilities :
1. Add a nativeElement property to AbstractControl, so the validator may check input validity
2. Add a property to AbstractControl that contains the text effectively entered by the user, or the validity of the input.
3. Change the operation of NumberValueAccessor so that it returns null if the user enters nothing, but returns NaN if user enters a non numeric string. This can be easily done by adding a condition in the registerOnChange function. After all, why is it that the control value should be different from what it contains ? And NumberValueAccessor will have a consistent behavior if later it had another selector that associate it with a input of type text (like input[datatype=number][formControlName]...)

**Comment by DzmitryShylovich** [comment]
Can u provide a repro in plunker?
> https://github.com/angular/angular/blob/master/modules/%40angular/forms/src/directives/number_value_accessor.ts#L51

`parseFloat` should return NaN fornon numeric string

**Comment by fredgate** [comment]
See my issue #11644 where I detail the problem.
Yes parseFloat return NaN for non numeric string. The problem is that NumberValueAccessor is apllied to input of type number, and that the value of these inputs is an empty string if the user entered a non numeric string.

**Comment by DzmitryShylovich** [comment]
> 1. Add a nativeElement property to AbstractControl, so the validator may check input validity
2. Add a property to AbstractControl that contains the text effectively entered by the user, or the validity of the input.

how it will solve your issue if according to https://github.com/angular/angular/issues/11644#issuecomment-247452664 a browser writes empty string as value of an element in case of non numeric string?

**Comment by fredgate** [comment]
In the NumberValueAccessor, we can access the validity state of the native element.
So there is 3 cases :
1. input is invalid : we return NaN
2. input is valid and value is empty string : we return null
3. input is valid and value is number : we return this number

**Comment by fredgate** [comment]
Or mey be we must keep a null value for the AbstractControl and the model in case of invalid input, and add a boolean property named `valid` to AbstractControl. It could avoid the NumberValueAccesor to manipulate the native element.

**Comment by evgd** [comment]
> parseFloat should return NaN for non numeric string
it produces unpredictable values for non numeric strings like
parseFloat('1x2') => 1
parseFloat('x1') => NaN
Number('1x2') => NaN - what you expect from number input

While in Chrome you can't enter invalid characters, in IE you can.
value for input of type "number" differs for IE. It seems IE doesn't perform proper sanitization and doesn't return empty string as value.
for "1x2" text in input:
IE 
value "1x2" 
valueAsNumber NaN

FF
value "" (empty string)
valueAsNumber NaN

Chrome
you can't enter characters, but there is way to put invalid number
for "22-33+11"
value "" (empty string)
valueAsNumber NaN

valueAsNumber for empty string all browsers returns NaN

it looks like valueAsNumber can be good cross-browser alternative to $event.target.value
`'(change)': 'onChange($event.target.value)',
    '(input)': 'onChange($event.target.value)',
`
or Number alternative to parseFloat in registerOnChange, but there might be false positives for Number("0b11") or Number("0xC"), but valueAsNumber even for IE will not pass this.
So far I suffer from missed validation in IE because of this.

**Comment by splincode** [comment]
@vsavkin Is the issue relevant?

**Comment by fredgate** [comment]
Soon 4 years, and we can not validate a number... :-(

**Comment by keatkeat87** [comment]
currently, what we can do is override NumberValueAccessor and use the input.validity to check valid or not. 
maybe like this : 
}
```
registerOnChange(fn: (_: number | null) => void): void {
  this.onChange = (e: Event) => {
    const input = e.target as HTMLInputElement;
    let value: number | null;
    if (input.validity.badInput) {
      value = NaN;
    } else if (input.value === '') {
      value = null;
    } else {
      value = parseFloat(input.value);
    }
    fn(value);
  };
}
```

**Comment by antch** [comment]
Similar to the solution provided by @keatkeat87, something like this also works in your own `ControlValueAccessor` component implementation:

```
  @ViewChild('myTemplateVar', {static: true})
  private inputEl: ElementRef;

  //snip
  registerOnChange(fn: any): void {
    this.onChangeCallback = (v: any): void => {
      const isNaN = v == null && this.inputEl.nativeElement.validity.badInput;
      fn(isNaN ? NaN : v);
    };
  }
```

**Comment by adam-marshall** [comment]
My problem is https://github.com/angular/angular/issues/11644. Any update on this? 

Another dimension to the problem is it will obviously block any attempts to use a custom validator to intercept. I have a pain point moving from AngularJS where the 'number' validation was built-in, and worked with ng-messages. I tried to add my own 'number' validator, but obviously this doesn't work as any non-numeric input never reaches my validator.

e,g,
```
static number(control: AbstractControl): ValidationErrors | null {
    if (!control.value) {
        return null;
    }

    const numberRegExp = /^\s*(\-|\+)?(\d+|(\d*(\.\d*)))\s*$/;

    if (numberRegExp.test(control.value)) {
        return null;
    }

    return { number: true };
}
```

**Comment by manklu** [comment]
@adam-marshall
see https://github.com/angular/angular/issues/2962#issuecomment-247549191 why you dont get a vlaue for numeric input fields

**Comment by adam-marshall** [comment]
@manklu yes thank you I had read the investigation which was undertaken by @fredgate in 2016.

I guess I am like the others in wondering in how this issue and others may now be prioritised in the wake of the release of Angular 9.

**Comment by amitdey32** [comment]
Is there a way to override the implementation of NumberValueAccessor.registerOnChange method in the application? Any idea is much appreciated.
